File: ant-sdk/antd-java/src/main/java/com/autonomi/antd/models/FinalizeUploadResult.java
package com.autonomi.antd.models;

/**
 * Result of finalizing an externally-signed upload.
 *
 * <p>{@code dataMap} and {@code dataMapAddress} were added in antd 0.6.1.
 * When prepare was called with {@code visibility="public"}, the DataMap chunk
 * is bundled into the same external-signer payment batch and stored on-network
 * — {@code dataMapAddress} is then the shareable retrieval handle. For
 * private/legacy uploads both fields are empty strings.
 *
 * @param address        legacy: set when {@code store_data_map=true} was passed
 *                       (paid by the daemon wallet). Empty otherwise.
 * @param chunksStored   number of chunks stored on the network
 * @param dataMap        hex-encoded serialized DataMap; always returned by
 *                       0.6.1+ daemons, empty when talking to older ones
 * @param dataMapAddress hex retrieval address of the DataMap chunk; set when
 *                       prepare was called with {@code visibility="public"},
 *                       empty otherwise
 */
public record FinalizeUploadResult(
        String address,
        long chunksStored,
        String dataMap,
        String dataMapAddress) {

    /**
     * Backward-compatible constructor for callers that only need the original
     * two fields. {@code dataMap} and {@code dataMapAddress} default to empty.
     */
    public FinalizeUploadResult(String address, long chunksStored) {
        this(address, chunksStored, "", "");
    }
}

File: ant-sdk/antd-php/src/Models/FinalizeUploadResult.php
<?php

declare(strict_types=1);

namespace Autonomi\Antd\Models;

/**
 * Result of finalizing an externally-signed upload.
 *
 * `dataMap` is always populated — it's the hex-encoded msgpack of the
 * underlying DataMap. `address` is only set when the legacy
 * `store_data_map=true` flag asked the daemon to pay+store the DataMap from
 * its own wallet (this PHP client doesn't expose that flag). `dataMapAddress`
 * is set when the prepare call used `visibility="public"` — in that case the
 * DataMap chunk was bundled into the same external-signer payment batch and
 * the returned address is the shareable retrieval handle. Defaults to "" when
 * talking to a pre-0.6.1 daemon that doesn't include the field.
 */
readonly class FinalizeUploadResult
{
    public function __construct(
        /** Hex-encoded msgpack DataMap (always returned). */
        public string $dataMap,
        /** Legacy: hex network address when store_data_map=true was passed (paid by daemon wallet). */
        public string $address,
        /** Hex network address when prepare used visibility="public" (paid in the external-signer batch). */
        public string $dataMapAddress,
        /** Number of chunks stored on the network. */
        public int $chunksStored,
    ) {}
}
